srcjs/init_shiny.js

```diff
@@ -82,24 +82,27 @@ function initShiny() {
   var inputsRate = new InputRateDecorator(inputsEvent);
   var inputsDefer = new InputDeferDecorator(inputsEvent);
 
-  // By default, use rate decorator
-  var inputs = inputsRate;
-  $('input[type="submit"], button[type="submit"]').each(function() {
+  var inputsValidate;
+  if ($('input[type="submit"], button[type="submit"]').length > 0) {
     // If there is a submit button on the page, use defer decorator
-    inputs = inputsDefer;
-    $(this).click(function(event) {
-      event.preventDefault();
-      inputsDefer.submit();
+    inputsValidate = new InputValidateDecorator(inputsDefer);
+
+    $('input[type="submit"], button[type="submit"]').each(function() {
+      $(this).click(function(event) {
+        event.preventDefault();
+        inputsDefer.submit();
+      });
     });
-  });
 
-  exports.onInputChange = function(name, value, opts) {
-    opts = Object.assign({
-      immediate: false,
-      binding: null,
-      el: null
-    }, opts);
+  } else {
+    // By default, use rate decorator
+    inputsValidate = new InputValidateDecorator(inputsRate);
+  }
 
+  var inputs = inputsValidate;
```

Comment:
Total nitpick but I would've had `var inputs` above the if/else block, and have `inputs = inputsRate` and `inputs = inputsDefer` as before, and then after the if/else block, `inputs = new InputValidateDecorator(inputs)`.